op-chain-ops/script/README.md

```diff
@@ -0,0 +1,139 @@
+host := NewHost(logger, artifactsFS, nil, DefaultContext, nil)
+```
+
+You can pass in optional arguments to `NewHost` to customize its behavior:
+
+- `WithBroadcastHook` will allow you to broadcast the transactions initiated using `vm.broadcast` on chain via an
+  external broadcaster.
+- `WithIsolatedBroadcast` will clear the warm state between calls, which allows broadcasted gas estimates to be more
+  accurate.
+- `WithCreate2Deployer` will configure the `CREATE2` opcode to broadcast transactions to the Arachnid `Create2Deployer`.
+
+Typically, Forge scripts in Go are implemented using the following pattern:
+
+1. Define a series of simple input and output contracts in Solidity.
+2. Implement your script, and have it take the input/output contracts as arguments to the `run` method.
+3. Define a struct that matches the input/output contracts in Go.
+4. Etch the input and output structs as contracts in Go, then call `run` using the etched input/output contract
+   addresses.
+5. Your script will run, and the output contract will contain the results.
+
+For an example, see [`opsm/implementations.go`](../deployer/opsm/implementations.go).
+
+# Known Limitations
+
+- Structs are not supported in contracts etched from Go code.
+- Bytecode containing linked libraries are not currently supported.
+
+# Best Practices
+
+## Keep I/O Contracts Simple
+
+I/O contracts are the bridge between the Go code and Solidity. You should keep them as simple as possible. Specifically:
+
+1. Use only primitive types in the I/O contracts. This means no structs, arrays, or mappings. The Go script engine will
+   automatically generate getters/setters for public struct fields.
+2. Avoid defining logic in your I/O contracts. Any logic you add in Solidity will have to be reimplemented in Go. If you
+   do add logic to your I/O contract (e.g., for validation) keep it simple.
+3. Avoid mutating your I/O contracts once they have been created.
+
+### Examples
+
+This contract:
+
+```solidity
+contract Input {
+    // Primitive types below
+    uint256 internal _a;
+    uint256 internal _b;
+    bytes internal _c;
+
+    // The below function will have to be reimplemented in Go, but it's really simple
+    function valuesSet() public view returns (bool) {
+        return _a != 0 && _b != 0 && _c.length != 0;
+    }
+}
+```
+
+Maps to this Go code:
+
+```go
+type Input struct {
+A uint64
+B uint64
+C []byte
+
+D uint64 `evm:"-"` // You can avoid creating setters/getters for this field using this tag
+}
+
+func (i *Input) ValuesSet() bool {
+return true
+}
+```
+
+The following are examples of antipatterns for I/O contracts:
+
+```solidity
+// Contracts with complex types
+contract Input {
+    MegaStruct internal _a; // structs are not supported
+}
+
+// Contracts containing logic that gets called from the script
+contract Input {
+    uint256 internal _a;
+
+    function setA(uint256 a) public {
+        require(a > 0, "a must be greater than 0");
+        _a = a;
+    }
+}
```

Comment:
I thought simple logic, like validations, was ok based on text above? Or is it ok, but still an antipattern, because it needs to be reimplemented?

We do this a lot currently, and on the solidity side if we don't want these checks in the input contracts we'd have to move them into the `run` method validations. So my current understanding is that we want all input and output checks to be in the script itself, not in the IO contracts